content/docs/user-guide/checkpoints.md

```diff
@@ -0,0 +1,195 @@
+
+## Adding checkpoints
+
+When you want to add a checkpoint to your training process, there are only a
+couple of things you need to setup. You can find the code we use here in
+[this GitHub repo](https://github.com/iterative/dvc-checkpoints-mnist).
+
+In your `dvc.yaml` file, add `checkpoint: true` to your `outs` under the model
+file. This adds checkpoints at the pipeline level. It should look something like
+this.
+
+```yaml
+stages:
+  train:
+    cmd: python train.py
+    deps:
+      - train.py
+    outs:
+      - model.pt:
+          checkpoint: true
+    live:
+      logs:
+        summary: true
+        html: true
+```
+
+You can also add checkpoints by running the command
+`dvc stage add -n name_of_stage --checkpoints model.pt`. This will create a new
+stage with checkpoints enabled.
+
+Then you can run your code using the `dvc exp run` command and it will start
+from an existing checkpoint or create a new checkpoint file for you. Now you
+have to include the checkpoints in your code using the `make_checkpoint()`
+function or the `DvcLiveCallback()` function.
+
+First, import the function into your script.
+
+`from dvc.api import make_checkpoint`
+
+You'll need a for-loop that has all of your training steps in and call
+`make_checkpoint` at the end. This does the same thing as running `dvc exp run`
+multiple times. You don't need to define a set number of epochs you want to run.
+The code will keep running and making checkpoints until you hit `Ctrl+C` to stop
+training. This will save the checkpoints that you have so far so you can take a
+look at them.
+
+While your script is executing epoch runs, you should see output in the terminal
+similar to this:
+
+```dvc
+Generating lock file 'dvc.lock'
+Updating lock file 'dvc.lock'
+Checkpoint experiment iteration 'fcefafb'.
+
+file:///Users/Repos/dvc-checkpoints-mnist/dvclive.html
+Updating lock file 'dvc.lock'
+Checkpoint experiment iteration '39d1444'.
+
+file:///Users/Repos/dvc-checkpoints-mnist/dvclive.html
+Updating lock file 'dvc.lock'
+Checkpoint experiment iteration '9b0bf16'.
+```
+
+If you execute `dvc exp show` after your epoch runs finish, you'll see a table
+with all of your checkpoints.
+
+```dvc
+┏━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━┳━━━━━━━━┓
+┃ Experiment    ┃ Created  ┃ step ┃    acc ┃
+┡━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━━━┩
+│ workspace     │ -        │    9 │ 0.9771 │
+│ live          │ 01:04 PM │    - │      - │
+│ │ ╓ exp-d3528 │ 01:21 PM │    9 │ 0.9771 │
+│ │ ╟ de0f0eb   │ 01:21 PM │    8 │  0.979 │
+│ │ ╟ 05e3cea   │ 01:21 PM │    7 │  0.974 │
+│ │ ╟ d0cd9ca   │ 01:20 PM │    6 │ 0.9708 │
+│ │ ╟ 0f495c4   │ 01:20 PM │    5 │ 0.9703 │
+│ │ ╟ 1131cba   │ 01:20 PM │    4 │ 0.9641 │
+│ │ ╟ 359869d   │ 01:20 PM │    3 │ 0.9521 │
+│ │ ╟ 9b0bf16   │ 01:20 PM │    2 │ 0.9271 │
+│ │ ╟ 39d1444   │ 01:19 PM │    1 │ 0.9157 │
+│ ├─╨ fcefafb   │ 01:19 PM │    0 │ 0.8572 │
+└───────────────┴──────────┴──────┴────────┘
+```
```

Comment:
It might be better to show an actual table that the user will get after the 1st run - the metrics will start from something like 0.10 in the 1st epoch, not 0.85.

Reply:
Also this table looks to be repeated in the **Viewing checkpoints** section. Probably just once is OK since there's another `exp show` example after `exp apply`?

Reply:
@dmpetrov The first iteration used to be a lot worse because it was using the full batch instead of minibatch. I primarily switched to minibatch to speed up the checkpoints, but it also made performance much better. Maybe in between would be good, since the old example took a bit too long to get decent results and tended to sometimes even get worse between checkpoints. The architecture is already pretty simple:

```
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
            Conv2d-1            [-1, 8, 28, 28]              80
         MaxPool2d-2            [-1, 8, 14, 14]               0
            Conv2d-3           [-1, 16, 14, 14]           1,168
            Linear-4                   [-1, 32]         100,384
            Linear-5                   [-1, 10]             330
================================================================
Total params: 101,962
Trainable params: 101,962
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.00
Forward/backward pass size (MB): 0.08
Params size (MB): 0.39
Estimated Total Size (MB): 0.48
----------------------------------------------------------------
```

We can probably make it even simpler (take out the second convolutional layer, reduce the number of channels even further) if you think it's "too easy" as is. Or we could switch to a different dataset (the recent dolt-dvc blog post had a good idea to use fashion-mnist as a harder alternative).

Reply:
@flippedcoder Maybe try out the different MNIST variants in https://pytorch.org/vision/stable/datasets.html? I think they all should be close to drop-in replacements but not as easy as the original MNIST.

Reply:
Sorry, I wasn't clear in my first response.

> the metrics will start from something like 0.10 in the 1st epoch, not 0.85.

What I meant is that this is not true now. The current repo jumps immediately to 0.85 after one epoch.